
syntax = "proto3";

package lemming.dataplane.sai;

import "dataplane/standalone/proto/common.proto";

option go_package = "github.com/openconfig/lemming/dataplane/standalone/proto";

enum PolicerAttr {
  POLICER_ATTR_UNSPECIFIED = 0;
  POLICER_ATTR_METER_TYPE = 1;
  POLICER_ATTR_MODE = 2;
  POLICER_ATTR_COLOR_SOURCE = 3;
  POLICER_ATTR_CBS = 4;
  POLICER_ATTR_CIR = 5;
  POLICER_ATTR_PBS = 6;
  POLICER_ATTR_PIR = 7;
  POLICER_ATTR_GREEN_PACKET_ACTION = 8;
  POLICER_ATTR_YELLOW_PACKET_ACTION = 9;
  POLICER_ATTR_RED_PACKET_ACTION = 10;
  POLICER_ATTR_ENABLE_COUNTER_PACKET_ACTION_LIST = 11;
}

message CreatePolicerRequest {
  option (sai_type) = OBJECT_TYPE_POLICER;
  uint64 switch = 1;
  optional MeterType meter_type = 2 [(attr_enum_value) = 1];
  optional PolicerMode mode = 3 [(attr_enum_value) = 2];
  optional PolicerColorSource color_source = 4 [(attr_enum_value) = 3];
  optional uint64 cbs = 5 [(attr_enum_value) = 4];
  optional uint64 cir = 6 [(attr_enum_value) = 5];
  optional uint64 pbs = 7 [(attr_enum_value) = 6];
  optional uint64 pir = 8 [(attr_enum_value) = 7];
  optional PacketAction green_packet_action = 9 [(attr_enum_value) = 8];
  optional PacketAction yellow_packet_action = 10 [(attr_enum_value) = 9];
  optional PacketAction red_packet_action = 11 [(attr_enum_value) = 10];
  repeated PacketAction enable_counter_packet_action_list = 12
      [(attr_enum_value) = 11];
}

message CreatePolicerResponse {
  uint64 oid = 1;
}

message RemovePolicerRequest {
  uint64 oid = 1;
}

message RemovePolicerResponse {}

message SetPolicerAttributeRequest {
  uint64 oid = 1;
  optional uint64 cbs = 2 [(attr_enum_value) = 4];
  optional uint64 cir = 3 [(attr_enum_value) = 5];
  optional uint64 pbs = 4 [(attr_enum_value) = 6];
  optional uint64 pir = 5 [(attr_enum_value) = 7];
  optional PacketAction green_packet_action = 6 [(attr_enum_value) = 8];
  optional PacketAction yellow_packet_action = 7 [(attr_enum_value) = 9];
  optional PacketAction red_packet_action = 8 [(attr_enum_value) = 10];
  repeated PacketAction enable_counter_packet_action_list = 9
      [(attr_enum_value) = 11];
}

message SetPolicerAttributeResponse {}

message GetPolicerAttributeRequest {
  uint64 oid = 1;
  repeated PolicerAttr attr_type = 2;
}

message GetPolicerAttributeResponse {
  PolicerAttribute attr = 1;
}

service Policer {
  rpc CreatePolicer(CreatePolicerRequest) returns (CreatePolicerResponse) {}
  rpc RemovePolicer(RemovePolicerRequest) returns (RemovePolicerResponse) {}
  rpc SetPolicerAttribute(SetPolicerAttributeRequest)
      returns (SetPolicerAttributeResponse) {}
  rpc GetPolicerAttribute(GetPolicerAttributeRequest)
      returns (GetPolicerAttributeResponse) {}
}
